### PR TITLE
Initial support for email query and get actions

### PR DESCRIPTION
**Pull request by jeserkin**

### CODE REVIEW

**Review by alagane — CHANGES_REQUESTED** [review]
Overall it looks good to me, I don't know yet when it will merged.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,11 @@
...
+/lib
+
+# IDEs and editors
+/.idea
+.project
+.classpath
+.c9/
+*.launch
+.settings/
+*.sublime-workspace
```
Comment (chibenwa) [inline .gitignore:11]: You can use a local, global ~/.gitignore for your own tool related stuff, a good practice is to put only build related ignores in a given project.

File: src/index.ts [hunk 2]
```diff
@@ -69,38 +79,20 @@ export class Client {
...
 
   public mailbox_get(
     args: IGetArguments<IMailboxProperties>
-  ): Promise<{
-    accountId: string | null;
-    state: string;
-    list: IMailbox[];
-    notFound: string;
-  }> {
+  ): Promise<IMailboxGetResponse> {
```
Comment (chibenwa) [inline src/index.ts:90]: :+1: for this strong typing! <3

**Review by chibenwa — COMMENTED** [review]
Thanks a lot for this very nice contribution! Those remarks are not related to this (nice!) PR but more to this repository code base in general.

Simple question but it would be really nice to use back-reference to chain JMAP calls and reduce server roundtrips.

(For someone like me based in Vietnam this is actually critical!)

You could:
 - start with a mailbox query to retrieve the guy INBOX
 - Then back-reference the INBOX in the email/query
 - And finally back-reference the ids in the Email/get

Doing all of that with a single network roundtrip, with a single API call.

Is there planned support for back-references in jmap-client-ts?

I would advise the maintainer to implement it early, in order to avoid generating technical debt.

-------------------------------

(No tests?)

I would suggest using linagora/james-memory docker image in order to start a JMAP server against which to test the client... :-)

File: .gitignore (hunk 1 above)
Comment (jeserkin) [reply .gitignore:11]: Sure. I can remove those.
Just as a separate question:
> Wouldn't that make life easier for other developers who would join development contributions in the future? So that they would not need to discover, that "Ah, my IDE sees some workspace related files, that I don't want to accidentally add to commit"

Comment (alagane) [reply .gitignore:11]: Well, I don't mind if there are some lines in the `.gitignore` since they is a comment to explain what they are related to.

**Review by alagane — APPROVED** [review]

File: src/types.ts [hunk 3]
```diff
@@ -74,6 +79,23 @@ export interface IEmail {
   createdModSeq: number;
   updatedModSeq: number;
   deleted: Date | null;
+  receivedAt: Date;
```
Comment (jeserkin) [inline src/types.ts:82]: I wonder if this and other `Date` fields should even be present as `Date` fields. You can't pass Date via JSON, it will be number or string, which end user will then need to convert. They look nice in type, but I feel like this is the wrong type for it. Library is intended to work with raw data. And raw data type for `Date` is `string`/`number`.

Comment (alagane) [reply src/types.ts:82]: Yes (here in the specs https://jmap.io/spec-core.html#the-date-and-utcdate-data-types)

We could use string or a custom type
```typescript
export type IDate = string;
```

Comment (jeserkin) [reply src/types.ts:82]: I think this ^ would work.

**Review by alagane — APPROVED** [review]

File: src/types.ts (hunk 3 above)
Comment (chibenwa) [reply src/types.ts:82]: JMAP mandat String ( date-time format [@!RFC3339]): https://jmap.io/spec-core.html#the-date-and-utcdate-data-types

**Comment by alagane** [comment]
`deleted` propery was in https://jmap.io/server.html#3-mailboxes but this page is not up to date (jmap draft) so some properties may be wrong.

File: src/types.ts [hunk 4]
```diff
@@ -52,28 +59,44 @@ export interface ISession {
...
+
+export interface IEmailKeywords {
+  $draft?: ITrue;
+  $seen?: ITrue;
+  $flagged?: ITrue;
+  $answered?: ITrue;
+  $forwarded?: ITrue;
+  $phishing?: ITrue;
+  $junk?: ITrue;
+  $notjunk?: ITrue;
```
Comment (chibenwa) [inline src/types.ts:94]: I'm not fluent in JavaScript but I got the feeling we only offer a close set of Keyworks.

JMAP allows passing arbitrary keywords. `Set<String>` is likely a better data structure to represent it (convenience methods can be exposed)

At least it can come in follow up refactorings.

File: src/types.ts [hunk 5]
```diff
@@ -52,28 +59,44 @@ export interface ISession {
...
+  from: IEmailAddress[] | null;
+  to: IEmailAddress[] | null;
   subject: string;
-  date: Date;
   size: number;
   preview: string;
   attachments: Attachment[] | null;
   createdModSeq: number;
   updatedModSeq: number;
-  deleted: Date | null;
```
Comment (chibenwa) [inline src/types.ts:76]: What are these createdModSeq updatedModSeq properties?

Looking at it in the spec I have 0 matches...

Is it necessary?

(I agree, this is not related to your work)

Comment (jeserkin) [reply src/types.ts:76]: I was wondering the same, but decided to lessen the overhead of changes. I was unable to locate those as well. However, they are present here https://jmap.io/server.html#3-mailboxes and as far as I understand, it is (draft aka not based on confirmed standard). Maybe @alagane can add some more detail in regards to it?

File: src/types.ts (hunk 4 above)
Comment (jeserkin) [reply src/types.ts:94]: Those keywords are related only to Email entity https://jmap.io/spec-mail.html#metadata as far as I can tell.

However:
> Users may add arbitrary keywords to an Email. For compatibility with IMAP, a keyword is a case-insensitive string of 1–255 characters in the ASCII subset %x21–%x7e (excludes control chars and space), and it MUST NOT include any of these characters

I would assume, that it could be added in future changes. Why I am offering it at the moment? Due to it still being in early stages it is easier to limit our logic to use of `interfaces`, `enums` and `types`. If we start to introduce classes it might become a bit more complex.

Due to JSON structure being returned the way it is (regarding keywords), we can't set it as `Set` without some conversion logic.

**Comment by jeserkin** [comment]
> `deleted` property was in https://jmap.io/server.html#3-mailboxes but this page is not up to date (jmap draft) so some properties may be wrong.

At the moment when I make Postman requests towards James I don't have this property in response. This is why I removed it at the moment.

**Comment by alagane** [comment]
I merged it, you can also create issues if needed, for example for things that are unrelated to the merge request but need to be done later.
In issues we can also discuss about some point that we would not agree on.
